amxmodx/file.cpp

```diff
@@ -7,484 +7,503 @@
+	int length;
+	const char* file = get_amxstring(amx, params[1], 0, length);
+
+	if (*params / sizeof(cell) >= 2 && params[2] > 0)
+	{
+		const char* pathID = get_amxstring(amx, params[3], 1, length);
+
+		if (!strcmp(pathID, "ALL"))
+		{
+			pathID = nullptr;
+		}
+
+		return ValveFile::Delete(file, pathID);
+	}
+
+	return SystemFile::Delete(build_pathname("%s", file));
 }
 
-static cell AMX_NATIVE_CALL file_exists(AMX *amx, cell *params) /* 1 param */
+// native file_exists(const file[], bool:use_valve_fs = false);
+static cell AMX_NATIVE_CALL file_exists(AMX *amx, cell *params)
 {
-	int iLen;
-	char *sFile = get_amxstring(amx, params[1], 0, iLen);
-	char *file = build_pathname("%s", sFile);
+	int length;
+	const char* file = get_amxstring(amx, params[1], 0, length);
 
-#if defined WIN32 || defined _WIN32
-	DWORD attr = GetFileAttributes(file);
-	
-	if (attr == INVALID_FILE_ATTRIBUTES)
-		return 0;
-	
-	if (attr == FILE_ATTRIBUTE_DIRECTORY)
-		return 0;
-	
-	return 1;
-#else
-	struct stat s;
-	
-	if (stat(file, &s) != 0)
-		return 0;
-	
-	if (S_ISDIR(s.st_mode))
-		return 0;
-	
-	return 1;
-#endif
+	if (*params / sizeof(cell) >= 2 && params[2] > 0)
+	{
+		return g_FileSystem->FileExists(file);
+	}
+
+	return g_LibSys.IsPathFile(build_pathname("%s", file));
 }
 
-static cell AMX_NATIVE_CALL dir_exists(AMX *amx, cell *params) /* 1 param */
+// native dir_exists(const dir[], bool:use_valve_fs = false);
+static cell AMX_NATIVE_CALL dir_exists(AMX *amx, cell *params)
 {
-	int iLen;
-	char *sFile = get_amxstring(amx, params[1], 0, iLen);
-	char *file = build_pathname("%s", sFile);
+	int length;
+	const char *path = get_amxstring(amx, params[1], 0, length);
+
+	if (*params / sizeof(cell) >= 2 && params[2] > 0)
+	{
+		return g_FileSystem->IsDirectory(path);
+	}
 
-	return DirExists(file) ? 1 : 0;
+	return g_LibSys.IsPathDirectory(build_pathname("%s", path));
 }
 
-static cell AMX_NATIVE_CALL file_size(AMX *amx, cell *params) /* 1 param */
+#define FSOPT_BYTES_COUNT  0
+#define FSOPT_LINES_COUNT  1
+#define FSOPT_END_WITH_LF  2
+
+// native file_size(const file[], flag = 0, bool:use_valve_fs = false, const valve_path_id[] = "GAME");
+static cell AMX_NATIVE_CALL file_size(AMX *amx, cell *params)
 {
-	int iLen;
-	char* sFile = get_amxstring(amx, params[1], 0, iLen);
-	AutoFilePtr fp(fopen(build_pathname("%s", sFile), "r"));
-	
-	if (fp != NULL)
+	int length;
+	const char* path = get_amxstring(amx, params[1], 0, length);
+	int flag = params[2];
+
+	AutoPtr<FileObject> fp;
+
+	size_t numParams = *params / sizeof(cell);
+
+	if (numParams >= 3 && params[3] > 0)
+	{
+		const char* pathID = get_amxstring(amx, params[4], 1, length);
+
+		if (!strcmp(pathID, "ALL"))
+		{
+			pathID = nullptr;
+		}
+
+		fp = ValveFile::Open(path, "r", pathID);
+	}
+	else
+	{
+		fp = SystemFile::Open(build_pathname("%s", path), "r");
+	}
+
+	if (!fp)
+	{
+		return -1;
+	}
+
+	switch (flag)
 	{
-		if (params[0] < 2 || params[2] == 0)
+		default:
```

Comment:
This shouldn't silently switch over to valid behavior, when it should actually have been an error path all along (I know we may not be able to fix that, but still).